usersession/client/client.go

```diff
@@ -0,0 +1,264 @@
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strconv"
+	"sync"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+)
+
+// dialSessionAgent connects to a user's session agent
+//
+// The host portion of the address is interpreted as the numeric user
+// ID of the target user.
+func dialSessionAgent(network, address string) (net.Conn, error) {
+	host, _, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	socket := filepath.Join(dirs.XdgRuntimeDirBase, host, "snapd-session-agent.socket")
+	return net.Dial("unix", socket)
+}
+
+type Client struct {
+	doer *http.Client
+}
+
+func New() *Client {
+	transport := &http.Transport{Dial: dialSessionAgent, DisableKeepAlives: true}
+	return &Client{
+		doer: &http.Client{Transport: transport},
+	}
+}
+
+type Error struct {
+	Kind    string      `json:"kind"`
+	Value   interface{} `json:"value"`
+	Message string      `json:"message"`
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+type response struct {
+	// Not from JSON
+	uid        int
+	err        error
+	statusCode int
+
+	Result json.RawMessage `json:"result"`
+	Type   string          `json:"type"`
+}
+
+func (resp *response) checkError() {
+	if resp.Type != "error" {
+		return
+	}
+	var resultErr Error
+	err := json.Unmarshal(resp.Result, &resultErr)
+	if err != nil || resultErr.Message == "" {
+		resp.err = fmt.Errorf("server error: %q", http.StatusText(resp.statusCode))
+	} else {
+		resp.err = &resultErr
+	}
+}
+
+func (client *Client) doMany(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body []byte) ([]*response, error) {
+	sockets, err := filepath.Glob(filepath.Join(dirs.XdgRuntimeDirGlob, "snapd-session-agent.socket"))
+	if err != nil {
+		return nil, err
+	}
+	var (
+		wg        sync.WaitGroup
+		mu        sync.Mutex
+		responses []*response
+	)
+	for _, socket := range sockets {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			uidStr := filepath.Base(filepath.Dir(socket))
+			uid, err := strconv.Atoi(uidStr)
+			if err != nil {
+				logger.Noticef("Socket %q does not appear to be in a valid XDG_RUNTIME_DIR", socket)
+				return
+			}
+			u := url.URL{
+				Scheme:   "http",
+				Host:     uidStr,
+				Path:     urlpath,
+				RawQuery: query.Encode(),
+			}
+			req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(body))
+			if err != nil {
+				logger.Noticef("Failed to create HTTP request: %v", err)
+				return
+			}
+			req = req.WithContext(ctx)
+			for key, value := range headers {
+				req.Header.Set(key, value)
+			}
+			httpResp, err := client.doer.Do(req)
+			response := response{
+				uid: uid,
+				err: err,
+			}
+			if err == nil {
+				defer httpResp.Body.Close()
+				response.statusCode = httpResp.StatusCode
+				response.err = decodeInto(httpResp.Body, &response)
+				response.checkError()
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			responses = append(responses, &response)
```

Comment:
I don't have a strong feeling about this, but I feel it would be slightly more idiomatic to use a channel instead of a waitgroup, a lock, and a list

Reply:
At the moment, there are a few error cases where the goroutine doesn't currently append to the slice: if there are garbage directories under `/run/user`, or if the http library can't create a request instance for some reason.

If we were using tomb, I suppose its `Dead` channel would be a way to know there are no more responses to read (since none of the goroutines are in a position to close the response channel), but that has its own cognitive overhead.